src/Functions/toStartOfInterval.cpp

```diff
@@ -40,156 +47,243 @@ class FunctionToStartOfInterval : public IFunction
-    Monotonicity getMonotonicityForRange(const IDataType &, const Field &, const Field &) const override
-    {
-        return { .is_monotonic = true, .is_always_monotonic = true };
-    }
+    Monotonicity getMonotonicityForRange(const IDataType &, const Field &, const Field &) const override { return { .is_monotonic = true, .is_always_monotonic = true }; }
 
     DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
     {
-        bool first_argument_is_date = false;
+        bool value_is_date = false;
         auto check_first_argument = [&]
         {
-            if (!isDate(arguments[0].type) && !isDateTime(arguments[0].type) && !isDateTime64(arguments[0].type))
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
-                    "Should be a date or a date with time", arguments[0].type->getName(), getName());
-            first_argument_is_date = isDate(arguments[0].type);
+            const DataTypePtr & type_arg1 = arguments[0].type;
+            if (!isDate(type_arg1) && !isDateTime(type_arg1) && !isDateTime64(type_arg1))
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of 1st argument of function {}. "
+                    "Should be a date or a date with time", type_arg1->getName(), getName());
+            value_is_date = isDate(type_arg1);
         };
 
         const DataTypeInterval * interval_type = nullptr;
-        bool result_type_is_date = false;
-        bool result_type_is_datetime = false;
-        bool result_type_is_datetime_64 = false;
-        auto check_interval_argument = [&]
+        enum class ResultType
         {
-            interval_type = checkAndGetDataType<DataTypeInterval>(arguments[1].type.get());
+            Date,
+            DateTime,
+            DateTime64
+        };
+        ResultType result_type;
+        auto check_second_argument = [&]
+        {
+            const DataTypePtr & type_arg2 = arguments[1].type;
+            interval_type = checkAndGetDataType<DataTypeInterval>(type_arg2.get());
             if (!interval_type)
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
-                    "Should be an interval of time", arguments[1].type->getName(), getName());
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of 2nd argument of function {}. "
+                    "Should be an interval of time", type_arg2->getName(), getName());
             switch (interval_type->getKind()) // NOLINT(bugprone-switch-missing-default-case)
             {
                 case IntervalKind::Nanosecond:
                 case IntervalKind::Microsecond:
                 case IntervalKind::Millisecond:
-                    result_type_is_datetime_64 = true;
+                    result_type = ResultType::DateTime64;
```

Comment:
There should be a comment between l. 81 and l. 82 which states that we determine the result type for the default overload.